examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,414 @@
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig, UncheckedFrom,
+};
+use sp_runtime::{
+    traits::{Saturating, StaticLookup, Zero},
+    DispatchError,
+};
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22BalanceOfInput<AssetId, AccountId> {
+    asset_id: AssetId,
+    owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22AllowanceInput<AssetId, AccountId> {
+    asset_id: AssetId,
+    owner: AccountId,
+    spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    to: AccountId,
+    value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    from: AccountId,
+    to: AccountId,
+    value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+    asset_id: AssetId,
+    spender: AccountId,
+    value: Balance,
+}
+
+#[derive(Default)]
+pub struct Psp22Extension;
+
+fn convert_err(err_msg: &'static str) -> impl FnOnce(DispatchError) -> DispatchError {
+    move |err| {
+        trace!(
+            target: "runtime",
+            "PSP22 Transfer failed:{:?}",
+            err
+        );
+        DispatchError::Other(err_msg)
+    }
+}
+
+enum FuncId {
+    Metadata(Metadata),
+    Query(Query),
+    Transfer,
+    TransferFrom,
+    Approve,
+    IncreaseAllowance,
+    DecreaseAllowance,
+}
+
+impl TryFrom<u16> for FuncId {
+    type Error = DispatchError;
+
+    fn try_from(func_id: u16) -> Result<Self, Self::Error> {
+        match func_id {
+            // Note: We use the first two bytes of PSP22 interface selectors as function IDs,
+            // While we can use anything here, it makes sense from a convention perspective.
+            0x3d26 => Ok(Self::Metadata(Metadata::Name)),
+            0x3420 => Ok(Self::Metadata(Metadata::Symbol)),
+            0x7271 => Ok(Self::Metadata(Metadata::Decimals)),
+            0x162d => Ok(Self::Query(Query::TotalSupply)),
+            0x6568 => Ok(Self::Query(Query::BalanceOf)),
+            0x4d47 => Ok(Self::Query(Query::Allowance)),
+            0xdb20 => Ok(Self::Transfer),
+            0x54b3 => Ok(Self::TransferFrom),
+            0xb20f => Ok(Self::Approve),
+            0x96d6 => Ok(Self::IncreaseAllowance),
+            0xfecb => Ok(Self::DecreaseAllowance),
+            _ => {
+                error!("Called an unregistered `func_id`: {:}", func_id);
+                return Err(DispatchError::Other("Unimplemented func_id"));
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Metadata {
+    Name,
+    Symbol,
+    Decimals,
+}
+
+#[derive(Debug)]
+enum Query {
+    TotalSupply,
+    BalanceOf,
+    Allowance,
+}
```

Comment:
Do you mind moving these above the `TryFrom` conversion and below the `FuncId` enum?